framework/include/base/Moose.h

```diff
@@ -54,6 +54,29 @@ class Factory;
 #define FORTRAN_CALL(name) name##_
 #endif
 
+/**
+ * Function to mirror the behavior of the C++17 std::map::try_emplace() method (no hint).
+ * @param m The std::map
+ * @param k The key use to insert the pair
+ * @param args The value to be inserted. This can be a moveable type but won't be moved
+ *             if the insertion is successful.
+ */
+template <class M, class... Args>
+std::pair<typename M::iterator, bool>
```

Comment:
Correct me if I'm wrong, but this looks like:

It'll work for std::map, which is great

It'll fail to compile with std::set and std::unordered_map, which is acceptable

And we can fix the latter situation later if needed by simply specializing more closely and adding set/unordered_map specializations alongside?

Reply:
I believe you are correct, and I also want to fix that... Maybe now rather than later. At the very least, I could start with a more limiting name `map_try_emplace` or `moose_map_try_emplace`.  If I was sufficiently talented, maybe I could get this to work with all three? Not sure if I have enough C++ foo to do that...

As Derek proposed, this really should go into libMesh eventually as it's super handy as you can see by the follow up commit.

Reply:
Any help on this would be appreciated. I really like being able to use the `libmesh_make_unique` macro. I want this to be as equally useful.

Reply:
Oh, definitely don't start with a more limiting name.  When you add specializations, you'll keep the existing name but just change the templating.  Instead of templating around "class M" you'll template around "class K, class V, class C, class A" and then explicitly use ```std::map<K,V,C,A>``` instead of ```M``` in the code.  Then you can add specializations for other containers, and the whole thing will still be API backwards compatible.